### ISSUE TITLE
[naming-convention] Allow underscore prefix for backing fields

### ISSUE THREAD

**Issue by glen-84** [typescript-eslint/typescript-eslint#816]
**Edit:** The description below referred to the `member-naming` rule, which has since been replaced by the new `naming-convention` rule. This issue is still relevant.

---

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/member-naming": ["error", { "private": "^(?!_)" }]
  }
}
```

```TS
class Example {
    private _var1; // Not a backing field, shouldn't start with an underscore.
    private _var2; // Allowed, as it's a backing field for the `var2` property.

    public get var2() {
        return this._var2;
    }
}
```

**Expected Result**

Error for `_var1`, but not for `_var2` (because it's a backing field).

**Actual Result**

Error for both fields.

**Additional Info**

Similar [TSLint issue](https://github.com/palantir/tslint/issues/1489) with 33 👍.

Since this rule is different to the one in TSLint, the implementation would be different. Maybe an `ignoreBackingFields: boolean` option, or a way to set the naming for backing fields separately.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.13.0` |
| `@typescript-eslint/parser`        | `1.13.0` |
| `TypeScript`                       | `3.4.5` |
| `ESLint`                           | `6.1.0` |
| `node`                             | `10.16.0` |
| `npm`                              | `6.9.0` |

**Comment by Josh-Cena** [comment]
Question: what counts as a backing field? A private field that only use in accessors? A private field that's returned from a getter? What about the following?

```js
class C {
  private _f1;
  private _f2;
  get f() {
    return this._f1 + this._f2;
  }
  set f(v) {
    this._f1 = v;
    this._f2 = v;
  }
}
```

...Or, should we keep the scope very limited and only detect getters that contain a single `return this._field`? That seems to defeat the purpose of getters though, which is to have additional gating logic.

**Comment by glen-84** [comment]
The initial idea was to only allow the underscore prefix if there was a getter with the exact same name (with or without actually checking that it returned that private field).

In your example, you wouldn't be able to prefix `f1` or `f2`, since there's no corresponding getter with those names, so no conflict.

If that is extended to "any private field that appears to be a backing field", then yes, it would need to check all accessors to ensure that at least one of them referenced the private field.

Different goals I guess. Perhaps there would need to be an option (any backing field, or only backing fields that would otherwise conflict with accessor names).

**Comment by Josh-Cena** [comment]
If it's only "private field with the same name as a getter", then I think it's easily doable. Sounds like a fair heuristic.

**Comment by JoshuaKGoldberg** [comment]
Per #8792, we're feature freezing this rule. It's reached the point of complexity where we're having a hard time justifying working on it. 

If you'd like to see this option happen, I'd encourage you to add it to a community project such as [eslint.style](https://eslint.style). Cheers! 💙

**Comment by karlhorky** [comment]
I also mentioned this just now over in my question to the ESLint Stylistic maintainers about potentially migrating this rule to ESLint Stylistic:

- https://github.com/typescript-eslint/typescript-eslint/issues/8792#issuecomment-2025344673
